From d790c74b39b632fb8518caf8b54583313f3de8d7 Mon Sep 17 00:00:00 2001
From: xuxiang <xuxiang@eswincomputing.com>
Date: Mon, 13 May 2024 20:40:38 +0800
Subject: [PATCH 027/109] WIN2030-15279:fix:power process bug

Changelogs:
1. init power add pmic power on,Solve the i2c function abnormality caused by the tri-state problem of som pin

Change-Id: I10678e2f437d2ea808ebf17c7626147fa4f071e7
---
 Core/Src/hf_common.c        |  1 +
 Core/Src/hf_i2c.c           |  9 +++++++--
 Core/Src/hf_power_process.c | 23 +++++++++++------------
 3 files changed, 19 insertions(+), 14 deletions(-)

diff --git a/Core/Src/hf_common.c b/Core/Src/hf_common.c
index d930eaf..0c6031a 100644
--- a/Core/Src/hf_common.c
+++ b/Core/Src/hf_common.c
@@ -1039,6 +1039,7 @@ uint32_t es_autoboot(void)
 	int32_t som_pwr_last_state = 0;
 	if(is_som_pwr_lost_resume() && !es_get_som_pwr_last_state(&som_pwr_last_state)) {
 		if (som_pwr_last_state){
+			printf("pwr enable and last state is power on\n");
 			return 1;
 		}
 	}
diff --git a/Core/Src/hf_i2c.c b/Core/Src/hf_i2c.c
index b388f75..ca79b58 100644
--- a/Core/Src/hf_i2c.c
+++ b/Core/Src/hf_i2c.c
@@ -26,8 +26,10 @@ int hf_i2c_reg_read(I2C_HandleTypeDef *hi2c, uint8_t slave_addr,
 	while (HAL_I2C_GetState(hi2c) != HAL_I2C_STATE_READY);
 	status = HAL_I2C_Mem_Read(hi2c, slave_addr, reg_addr, I2C_MEMADD_SIZE_8BIT,
 							  data_ptr, 0x1, 0xff);
-	if (status != HAL_OK)
+	if (status != HAL_OK){
 		printf("I2Cx_read_Error(%x) reg %x; status %x\r\n", slave_addr, reg_addr, status);
+		return status;
+	}
 	while (HAL_I2C_GetState(hi2c) != HAL_I2C_STATE_READY);
 	while (HAL_I2C_IsDeviceReady(hi2c, slave_addr, 0xff, 0xff) == HAL_TIMEOUT);
 	while (HAL_I2C_GetState(hi2c) != HAL_I2C_STATE_READY);
@@ -43,7 +45,10 @@ int hf_i2c_mem_read(I2C_HandleTypeDef *hi2c, uint8_t slave_addr,
 	status = HAL_I2C_Mem_Read(hi2c, slave_addr, reg_addr, I2C_MEMADD_SIZE_8BIT,
 					data_ptr, len, 1000);
 	if (status != HAL_OK)
-		printf("I2Cx_read_Error(%x);\r\n", slave_addr);
+	{
+		printf("I2Cx_read_Error(%x) reg %x; status %x\r\n", slave_addr, reg_addr, status);
+		return status;
+	}
 	while (HAL_I2C_GetState(hi2c) != HAL_I2C_STATE_READY);
 	while (HAL_I2C_IsDeviceReady(hi2c, slave_addr, 0xff, 0xff) == HAL_TIMEOUT);
 	return status;
diff --git a/Core/Src/hf_power_process.c b/Core/Src/hf_power_process.c
index c21545d..e9318a2 100644
--- a/Core/Src/hf_power_process.c
+++ b/Core/Src/hf_power_process.c
@@ -124,7 +124,6 @@ void hf_power_task(void *parameter)
 #else
 void hf_power_task(void *parameter)
 {
-	HAL_StatusTypeDef status = HAL_OK;
 	power_state_t power_state = IDLE_STATE;
 	GPIO_PinState pin_state = GPIO_PIN_RESET;
 	printf("hf_power_task started!!!\r\n");
@@ -145,17 +144,9 @@ void hf_power_task(void *parameter)
 		case DC_PWR_GOOD_STATE:
 			printf("DC_PWR_GOOD_STATE\r\n");
 			power_state = SOM_STATUS_CHECK_STATE;
-			i2c_init(I2C3);
 			break;
 		case SOM_STATUS_CHECK_STATE:
 			printf("SOM_STATUS_CHECK_STATE\r\n");
-			pmic_power_on(pdTRUE);
-			do
-			{
-				osDelay(200);
-				status = pmic_b6out_105v();
-				osDelay(50);
-			}while(status != pdTRUE);
 			som_reset_control(pdFALSE);
 			pmic_status_led_on(pdTRUE);
 			power_led_on(pdTRUE);
@@ -190,6 +181,7 @@ void hf_power_task(void *parameter)
 void power_test_dome(void)
 {
 	#ifdef POWER_TESET_MODE
+	HAL_StatusTypeDef status = HAL_OK;
 	uint8_t pin_state = 0;
 	printf("atx_power_on\r\n");
 	atx_power_on(pdTRUE);
@@ -200,6 +192,14 @@ void power_test_dome(void)
 	} while (pin_state != pdTRUE);
 	printf("dc_power good\r\n");
 	osDelay(100);
+	i2c_init(I2C3);
+	pmic_power_on(pdTRUE);
+	do
+	{
+		osDelay(200);
+		status = pmic_b6out_105v();
+		osDelay(50);
+	}while(status != pdTRUE);
 	i2c_init(I2C1);
 	#endif
 }
@@ -207,12 +207,11 @@ void power_test_dome(void)
 void set_power_off(void)
 {
 	i2c_deinit(I2C3);
+	som_reset_control(pdTRUE);
 	#ifndef POWER_TESET_MODE
 	i2c_deinit(I2C1); // dvb v2 move to board init
-	#endif
 	pmic_power_on(pdFALSE);
 	osDelay(10);
-	#ifndef POWER_TESET_MODE
 	atx_power_on(pdFALSE);
 	osDelay(10);
 	#endif
@@ -300,12 +299,12 @@ void som_reset_control(uint8_t reset)
 	GPIO_InitTypeDef GPIO_InitStruct = {0};
 	int dip_switch_soft_ctl_attr = 0, dip_switch_soft_state = 0;
 	if (reset) {
+		HAL_GPIO_WritePin(MCU_RESET_SOM_N_GPIO_Port, MCU_RESET_SOM_N_Pin, GPIO_PIN_RESET);
 		GPIO_InitStruct.Pin = MCU_RESET_SOM_N_Pin;
 		GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
 		GPIO_InitStruct.Pull = GPIO_NOPULL;
 		GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
 		HAL_GPIO_Init(MCU_RESET_SOM_N_GPIO_Port, &GPIO_InitStruct);
-		HAL_GPIO_WritePin(MCU_RESET_SOM_N_GPIO_Port, MCU_RESET_SOM_N_Pin, GPIO_PIN_RESET);
 		uart_deinit(UART4);
 		uart_deinit(USART6);
 		SPI2_FLASH_CS_LOW();
-- 
2.25.1

